app/uk/gov/hmrc/cgtpropertydisposalsfrontend/services/audit/AuditService.scala

```diff
@@ -0,0 +1,607 @@
+    oldContactAddress: Address,
+    newContactAddress: Address,
+    isManuallyEnteredAddress: Boolean,
+    path: String
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Unit = {
+
+    val source = if (isManuallyEnteredAddress) "manual" else "postcode-lookup"
+
+    val details: RegistrationContactAddressChangedEvent = RegistrationContactAddressChangedEvent(
+      oldContactAddress,
+      newContactAddress,
+      source
+    )
+
+    sendEvent(
+      "cgt-property-disposals",
+      "registrationContactAddressChanged",
+      Json.toJson(details),
+      hc.toAuditTags("registration-contact-address-changed", path)
+    )
+  }
+
+  override def sendRegistrationChangeEmailAddressAttemptedEvent(
+    oldEmailAddress: String,
+    newEmailAddress: String,
+    path: String
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Unit = {
+
+    val detail = RegistrationChangeEmailAttemptedEvent(
+      oldEmailAddress,
+      newEmailAddress
+    )
+
+    sendEvent(
+      "cgt-property-disposals",
+      "registrationChangeEmailAddressAttempted",
+      Json.toJson(detail),
+      hc.toAuditTags("registration-change-email-address-attempted", path)
+    )
+  }
+
+  override def sendRegistrationChangeEmailVerifiedEvent(
+    oldEmailAddress: String,
+    newEmailAddress: String,
+    path: String
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Unit = {
+    val detail = RegistrationChangeEmailAddressVerifiedEvent(
+      oldEmailAddress,
+      newEmailAddress
+    )
+
+    sendEvent(
+      "cgt-property-disposals",
+      "registrationChangeEmailAddressVerified",
+      Json.toJson(detail),
+      hc.toAuditTags("registration-change-email-address-verified", path)
+    )
+  }
+
+  override def sendRegistrationSetupEmailAttemptedEvent(
+    emailAddress: String,
+    path: String
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Unit = {
+    val detail = RegistrationSetupEmailAttemptedEvent(
+      emailAddress
+    )
+
+    sendEvent(
+      "cgt-property-disposals",
+      "registrationSetupEmailAddressAttempted",
+      Json.toJson(detail),
+      hc.toAuditTags("registration-setup-email-address-attempted", path)
+    )
+  }
+
+  override def sendRegistrationSetupEmailVerifiedEvent(
+    emailAddress: String,
+    path: String
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Unit = {
+    val detail = RegistrationSetupEmailVerifiedEvent(
+      emailAddress
+    )
+
+    sendEvent(
+      "cgt-property-disposals",
+      "registrationSetupEmailAddressVerified",
+      Json.toJson(detail),
+      hc.toAuditTags("registration-setup-email-address-verified", path)
+    )
+  }
+
+  override def sendRegistrationRequestEvent(
+    registrationDetails: RegistrationDetails,
+    path: String
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Unit = {
+
+    val detail = RegistrationRequestEvent(
+      RegistrationPrePopulatedUserData(
+        "CGT",
+        EmailAuditDetails(
```

Comment:
the email address can either be prepopulated or manually entered but not both - it's prepopulated if there's a gg email and the user hasn't changed it. Otherwise, if there's no gg email then it'll be manual or if there is a gg email but the user has changed it from the check your details screen then it is manual as well.

Reply:
fixed